### PR TITLE
Improving optimizer performance by eliminating unnecessary sort and distribution requirements

### PR DESCRIPTION
**Pull request by metesynnada**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/5715.

# Rationale for this change

The current implementation of SymmetricHashJoin requires order information before the PipelineFixer. This limitation results in unnecessary sort and distribution enforcement requirements that impact the optimizer's performance. To overcome this issue, we have revamped `SymmetricHashJoin` to eliminate the need for order information before PipelineFixer.

# What changes are included in this PR?

We have modified SymmetricHashJoin to function without requiring order information. The new implementation does not raise errors without order or filter information, though pruning is not supported without order information. Furthermore, we have set the required_input_ordering API to None, enabling the source to provide order information for piped executions. With this approach, we can genuinely remove the dependency on EnforceSorting and EnforceDistribution from PipelineFixer.

# Are these changes tested?

Yes.

# Are there any user-facing changes?

A new configuration option (`allow_symmetric_joins_without_pruning`) allows the user to constrain `SymmetricHashJoin`s to run only on ordered inputs—no breaking changes in existing use cases.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/pipeline_fixer.rs [hunk 1]
```diff
@@ -110,42 +104,17 @@ fn check_support(expr: &Arc<dyn PhysicalExpr>) -> bool {
...
-                return Ok(expr_supported
-                    && fields_supported
-                    && left_convertible
-                    && right_convertible);
-            }
-        }
-    }
-    Ok(false)
+fn is_suitable_for_symmetric_hash_join(hash_join: &HashJoinExec) -> bool {
+    hash_join.filter().map_or(false, |filter| {
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/pipeline_fixer.rs:110]: Should the default value here be `true` now? We are allowing SHJ to run even if the filter is not on sorted inputs, so why not do the same when there is no filter?

Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/pipeline_fixer.rs:110]: Should the default value here be `true` now? We are allowing SHJ to run even if the filter is not on sorted inputs, so why not do the same when there is no filter?

**Review by ozankabak — COMMENTED** [review]
Sent a commit with some simplifications and left one in-line comment, updated PR body to mention the new configuration option.

Things you must do before sending upstream:
1. Finding a solution the problem we discussed yesterday.
2. Getting the `Option` handling in SHJ to a minimal level. I already tried to help a little bit here, but most of the work is still outstanding.

**Comment by ozankabak** [comment]
I made some changes regarding SHJ suitability checks and did some code simplifications in SHJ code. Let's have a short meeting to go over the changes, and we can send this upstream after that if no issue emerges.

**Comment by ozankabak** [comment]
@metesynnada, AFAICT the SHJ test running too long issue intermittently happens here too (take a look at the failing test). So maybe it is a previously existing issue?

**Comment by metesynnada** [comment]
@ozankabak After we get rid of locks, the issue is resolved (hopefully), if you have any suggestion, I can try.

**Comment by ozankabak** [comment]
Merged upstream.

**Pull request closed without merging**